#Task 4.5_Data Consistency Checks

Importing libraries

In [5]:
# Import libraries
import pandas as pd
import numpy as np
import os

Importing data

In [6]:
#Create project folder path
path=r'C:\Users\Bori\Documents\Careerfoundry\Achievement 4'

In [7]:
#Import the “orders_wrangled.csv” file into Jupyter as df_ords
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'))

In [8]:
#Import the “products.csv” file into Jupyter as df_prods
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'))

Mixed type data

In [9]:
#Create a data frame
df_test = pd.DataFrame()
#Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [10]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [11]:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


Changing data type to string

In [12]:
df_test['mix'] = df_test['mix'].astype('str')

Checking for missing values

In [13]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

Creating a subset of missing values

In [14]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [15]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

Checking shape of data frame

In [16]:
df_prods.shape

(49693, 5)

Addressing missing values

In [17]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

Checking shape of data frame without missing values

In [18]:
df_prods_clean.shape

(49677, 5)

Checking for duplicates

In [19]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

Addressing duplicates

In [20]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [21]:
df_prods_clean_no_dups.shape

(49672, 5)

In [22]:
df_prods_clean_no_dups

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49672 rows x 5 columns]

#Exercise 4.5

If you haven’t performed the consistency checks covered in this Exercise on your df_prods dataframe, do so now

In [25]:
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

Run the df.describe() function on your df_ords dataframe.

In [26]:
df_ords.describe()

Unnamed: 0      order_id       user_id  prior_orders_by_user  \
count  3.421083e+06  3.421083e+06  3.421083e+06          3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.710542e+06          1.715486e+01   
std    9.875817e+05  9.875817e+05  9.875817e+05          1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00          1.000000e+00   
25%    8.552705e+05  8.552715e+05  8.552715e+05          5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.710542e+06          1.100000e+01   
75%    2.565812e+06  2.565812e+06  2.565812e+06          2.300000e+01   
max    3.421082e+06  3.421083e+06  3.421083e+06          1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

Order_id and user_id data type probably should be converted to string, calculating their mean, etc. simply doesn't make sense. The prior_orders_by_user and days_since_prior_order have some unexpected results, in both case, maximum is lower number that quartile or median. This should be further examined. Also, order_hour_of_day column contains some strange results. It should be integer between 0-12 or 0-24. I would like to examine it furter. Column orders_day_of_week seems ok, it should be a number between 1-7 (referring to the days of the week, respectively), 0-6 it seems ok also.

Check for mixed-type data in your df_ords dataframe

In [27]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col, 'mixed')
  else: print (col, ' consistent')

Unnamed: 0  consistent
order_id  consistent
user_id  consistent
eval_set  consistent
prior_orders_by_user  consistent
orders_day_of_week  consistent
order_hour_of_day  consistent
days_since_prior_order  consistent


If you find mixed-type data, fix it. The column in question should contain observations of a single data type.
All collumns have observations of a a single data type, no additional steps are needed.

Run a check for missing values in your df_ords dataframe

In [28]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
prior_orders_by_user           0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

Only column days_since_prior_order contains missing values. This column contains a lot of missing values, 206209 observations are missing.

Addressing missing values

In [29]:
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [30]:
df_ords_nan

Unnamed: 0  order_id  user_id eval_set  prior_orders_by_user  \
0                 0   2539329  2539329    prior                     1   
11               11   2168274  2168274    prior                     1   
26               26   1374495  1374495    prior                     1   
39               39   3343014  3343014    prior                     1   
45               45   2717275  2717275    prior                     1   
...             ...       ...      ...      ...                   ...   
3420930     3420930    969311   969311    prior                     1   
3420934     3420934   3189322  3189322    prior                     1   
3421002     3421002   2166133  2166133    prior                     1   
3421019     3421019   2227043  2227043    prior                     1   
3421069     3421069   3154581  3154581    prior                     1   

         orders_day_of_week  order_hour_of_day  days_since_prior_order  
0                         2                  8                     NaN  
11                        2                 11                     NaN  
26                        1                 14                     NaN  
39                        6                 11                     NaN  
45                        3                 12                     NaN  
...                     ...                ...                     ...  
3420930                   4                 12                     NaN  
3420934                   3                 18                     NaN  
3421002                   6                 19                     NaN  
3421019                   1                 15                     NaN  
3421069                   3                 11                     NaN  

[206209 rows x 8 columns]

It seems most users with missing values have 1 in the prior order column. Maybe it means it is the first order of the user, therefore days_since_prior_order column doesn't make sense in their case, since there was no prior order. 

In [31]:
df_ords[df_ords['prior_orders_by_user'] == 1]

Unnamed: 0  order_id  user_id eval_set  prior_orders_by_user  \
0                 0   2539329  2539329    prior                     1   
11               11   2168274  2168274    prior                     1   
26               26   1374495  1374495    prior                     1   
39               39   3343014  3343014    prior                     1   
45               45   2717275  2717275    prior                     1   
...             ...       ...      ...      ...                   ...   
3420930     3420930    969311   969311    prior                     1   
3420934     3420934   3189322  3189322    prior                     1   
3421002     3421002   2166133  2166133    prior                     1   
3421019     3421019   2227043  2227043    prior                     1   
3421069     3421069   3154581  3154581    prior                     1   

         orders_day_of_week  order_hour_of_day  days_since_prior_order  
0                         2                  8                     NaN  
11                        2                 11                     NaN  
26                        1                 14                     NaN  
39                        6                 11                     NaN  
45                        3                 12                     NaN  
...                     ...                ...                     ...  
3420930                   4                 12                     NaN  
3420934                   3                 18                     NaN  
3421002                   6                 19                     NaN  
3421019                   1                 15                     NaN  
3421069                   3                 11                     NaN  

[206209 rows x 8 columns]

In [32]:
df_ords[df_ords['prior_orders_by_user'] == 2]

Unnamed: 0  order_id  user_id eval_set  prior_orders_by_user  \
1                 1   2398795  2398795    prior                     2   
12               12   1501582  1501582    prior                     2   
27               27    444309   444309    prior                     2   
40               40   2030307  2030307    prior                     2   
46               46   1909121  1909121    prior                     2   
...             ...       ...      ...      ...                   ...   
3420931     3420931   2658896  2658896    prior                     2   
3420935     3420935   3351137  3351137    prior                     2   
3421003     3421003   1074448  1074448    prior                     2   
3421020     3421020   1959749  1959749    prior                     2   
3421070     3421070   1889163  1889163    prior                     2   

         orders_day_of_week  order_hour_of_day  days_since_prior_order  
1                         3                  7                    15.0  
12                        5                 10                    10.0  
27                        3                 19                     9.0  
40                        4                 11                    19.0  
46                        0                 16                    11.0  
...                     ...                ...                     ...  
3420931                   2                 15                    30.0  
3420935                   6                 17                     3.0  
3421003                   0                 10                     1.0  
3421020                   2                 14                     8.0  
3421070                   3                 17                     7.0  

[206209 rows x 8 columns]

In [33]:
df_ords[df_ords['prior_orders_by_user'] == 3]

Unnamed: 0  order_id  user_id eval_set  prior_orders_by_user  \
2                 2    473747   473747    prior                     3   
13               13   1901567  1901567    prior                     3   
28               28   3002854  3002854    prior                     3   
41               41    691089   691089    prior                     3   
47               47   2267326  2267326    prior                     3   
...             ...       ...      ...      ...                   ...   
3420932     3420932    414137   414137    prior                     3   
3420936     3420936    540764   540764    prior                     3   
3421004     3421004   1666077  1666077    prior                     3   
3421021     3421021   3392839  3392839    prior                     3   
3421071     3421071   1542354  1542354    prior                     3   

         orders_day_of_week  order_hour_of_day  days_since_prior_order  
2                         3                 12                    21.0  
13                        1                 10                     3.0  
28                        3                 16                    21.0  
41                        4                 15                    21.0  
47                        3                 18                    10.0  
...                     ...                ...                     ...  
3420932                   5                 16                    10.0  
3420936                   6                 19                     7.0  
3421004                   2                 15                    23.0  
3421021                   2                 15                     7.0  
3421071                   5                 11                    30.0  

[206209 rows x 8 columns]

Through the three above examples  I can tell that any user who has an order number of \"1\" will have \"NaN\" as their days_since_prior_order. This make sense, since it was their first order, days since prior order cannot be calculated. Maximum user id is 206209 and total missing values for days_since_prior_order is also 206209, meaning each user will have one NaN value located in their first order. I decided not to make any changes. Even though it looks missing, actually the data itself has value in it. We can use it to show that it's actually their first order of the respective user.

Run a check for duplicate values in your df_ords data.

In [34]:
df_ords_dups = df_ords[df_ords.duplicated()]

In [35]:
df_ords_clean = df_ords.drop_duplicates()

In [36]:
df_ords_clean.shape

(3421083, 8)

In [37]:
df_ords.shape

(3421083, 8)

No duplicates were found.

Export df_prods_clean_no_dups

In [38]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

In [39]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))